PLUGIN-DESIGN.md

```diff
@@ -0,0 +1,376 @@
+-   Transform the contents of a file before consuming it
+
+## Project file changes
+
+Add a new top-level field to the project file format:
+
+-   `plugins`: An array of [Plugin Descriptions](#plugin-description).
+    -   **Optional**
+    -   Default is `[]`
+
+### Plugin description
+
+Either a `String` value or an object with the fields:
+
+-   `source`: A filepath to the Lua source file of the plugin or a URL to a GitHub repo, optionally
+    followed by an `@` character and a Git tag.
+    -   **Required**
+-   `options`: Any JSON dictionary. The options that will be passed to the plugin.
+    -   **Optional**
+    -   Default is `{}`
+
+In the case that the value is just a `String`, it is interpreted as an object with the `source`
+field set to its value, and `options` set to its default.
+
+### Example
+
+```json
+{
+    "name": "ProjectWithPlugins",
+    "plugins": [
+        "local-plugin.lua",
+        "github.com/owner/remote-plugin-from-tag@v1.0.0",
+        "github.com/owner/remote-plugin-from-head",
+        { "source": "plugin-with-options.lua", "options": { "some": "option" } }
+    ],
+    ...
+}
+```
+
+## Plugin scripts
+
+Plugin scripts should return a `CreatePluginFunction`:
+
+```luau
+-- Types provided in luau format
+
+type PluginInstance = {
+  name: string,
+  projectDescription?: (project: ProjectDescription) -> (),
+  syncStart?: () -> (),
+  syncEnd?: () -> (),
+  resolve?: (id: string) -> string,
+  middleware?: (id: string) -> string,
+  load?: (id: string) -> string,
+}
+
+-- TODO: Define properly. This is basically just the JSON converted to Lua
+type ProjectDescription = { ... }
+
+type CreatePluginFunction = (options: {[string]: any}) -> PluginInstance
+```
+
+In this way, plugins have the opportunity to customize their hooks based on the options provided by
+the user in the project file.
+
+## Plugin instance
+
+-   `name`
+    -   **Required**: The name of the plugin that will be used in error messages, etc.
+-   `projectDescription(project: ProjectDescription) -> ()`
+    -   **Optional**: Called with a Lua representation of the current project description whenever
+        it has changed.
+-   `syncStart() -> ()`
+    -   **Optional**: A sync has started.
+-   `syncEnd() -> ()`
+    -   **Optional**: A sync has finished.
+-   `resolve(id: string) -> string`
+    -   **Optional**: Takes a file path and returns a new file path that the file should be loaded
+        from instead. The first plugin to return a non-nil value per id wins.
+-   `middleware(id: string) -> string`
+    -   **Optional**: Takes a file path and returns a snapshot middleware enum to determine how Rojo
+        should build the instance tree for the file. The first plugin to return a non-nil value per
+        id wins.
+-   `load(id: string) -> string`
+    -   **Optional**: Takes a file path and returns the file contents that should be interpreted by
+        Rojo. The first plugin to return a non-nil value per id wins.
```

Comment:
Having a quick read over this doc, making plugins race against each other seems a bit odd, and could lead to some nondeterminism in different runs (plugin 1 was quick the first time round, but something happened and it was slow the second time, so its output is now lost).

Maybe it would be nicer to have a specific ordering of the plugins? the `plugins` field in the project file format even leads to this - a user can order the plugins in the array depending on how they want them to run. (e.g. we have plugin 1 which firstly transforms some code, then we let plugin 2 run afterwards which takes the output from plugin 1, and minifies it)

Reply:
I'm not quite sure where you got the idea that the plugins would race each other but your suggestion is actually how it's already intended to work. The `plugins` array allows the user to determine the plugin order and that order is used each time to deterministically determine the results.

Reply:
The proposal said "the first plugin to return a non nil value wins" but maybe I misunderstood this - I thought it meant all plugins run at the same time and the first one to return a value will be used. I now realise I think it means that once a value is returned, it no longer calls for values from other plugins? My apologies!

Still, I think the second point mentioned may still make sense. If plugin 1 transforms require paths, and plugin 2 minifies the whole code, then we want to run *both* plugins (plugin 1 then 2), rather than just stopping after 1. I haven't looked into implementation details though so if this actually is what is happening already then my bad.

Reply:
Ahh I understand the confusion now. Yes, that sentence means "run them in order and stop once one has returned a value."

Yes, that is a good point! I will need to get my head back into how this proposal works (it's been a while) and come back with a response on it but that's definitely a use case that should be supported.